### PR TITLE
Add telemetry for sign out

### PR DESCRIPTION
**Pull request by 928PJY**
split from #35
Implement https://ceapex.visualstudio.com/Engineering/_workitems/edit/171972

### CODE REVIEW

File: src/observers/telemetryObserver.ts [hunk 1]
```diff
@@ -16,6 +16,12 @@ export class TelemetryObserver {
             case EventType.UserSignInCompleted:
                 this.handleUserSignInCompleted(<UserSignInCompleted>event);
                 break;
+            case EventType.UserSignOutTriggered:
+                this.handleUserSignOutTriggered(<UserSignOutTriggered>event);
+                break;
+            case EventType.UserSignOutCompleted:
```
Comment (live1206) [inline src/observers/telemetryObserver.ts:22]: Since there is no global config to route the event to which observer, and this `UserSignOutCompleted` has been handled in many observers, just wondering is there a way to send an event to a specific handler, not broadcasting? Should that be more efficient, not all handlers would go through all the events? This question is not related to this PR.

File: src/observers/signStatusBarObserver.ts [hunk 2]
```diff
@@ -28,7 +28,6 @@ export class SignStatusBarObserver extends BaseStatusBarObserver {
                 let asCredentialRetrieveFromLocalCredentialManager = <CredentialRetrieveFromLocalCredentialManager>event;
                 this.handleSignedIn(asCredentialRetrieveFromLocalCredentialManager.credential);
                 break;
-            case EventType.UserSignedOut:
```
Comment (live1206) [inline src/observers/signStatusBarObserver.ts:31]: no need to refresh the status bar for `UserSignedOut`?

File: src/observers/telemetryObserver.ts (hunk 1 above)
Comment (928PJY) [reply src/observers/telemetryObserver.ts:22]: For your question, the answer is no. that would be more efficient if we send the event to a specific handler(actually if we do, we even don't need to use the event-driven framework, we can directly call it as a function), but that would lose the advantage that the event sender does not need to care about the event consumer(decouple the sender and consumer), which is the principle of event-driven.

File: src/observers/signStatusBarObserver.ts (hunk 2 above)
Comment (928PJY) [reply src/observers/signStatusBarObserver.ts:31]: when the user sign out, the `resetCredential` function will be called, and a `CredentialReset` event will be post, which will trigger the statusbar refresh

**Review by live1206 — APPROVED** [review]

File: src/observers/telemetryObserver.ts (hunk 1 above)
Comment (live1206) [reply src/observers/telemetryObserver.ts:22]: OK, I see your point.
But do we always need to follow this sender and consumer decoupling principle?
For instance within micro-services, the component needs to send messages to the other components(here comes two ways: send or broadcast) instead of function call since they are not in the same application.
Is our extension that complex so we want to avoid function call?

Comment (live1206) [reply src/observers/telemetryObserver.ts:22]: I mean, we do have function calls.
Then what is the principle to use the event-driven way, like invoking between UI/logging/telemetry and extension main component?

Comment (yufeih) [reply src/observers/telemetryObserver.ts:22]: I second the use of simple function calls. Just look at the amount of code needed to log a simple telemetry or update a simple piece of text.
